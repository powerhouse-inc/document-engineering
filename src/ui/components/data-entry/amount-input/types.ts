import type { Currency } from '../currency-code-picker/types.js'

export interface Amount {
  value?: number
  unit?: CurrencyTicker
}

export type AmountPercentage = number | undefined
export type CurrencyTicker = Currency['ticker']

export interface AmountFiat {
  value?: number
  unit: CurrencyTicker
}

export interface AmountCrypto {
  value?: string
  unit: CurrencyTicker
}

export interface AmountCurrency {
  value?: string
  unit: CurrencyTicker
}

export type AmountInputPropsGeneric =
  | {
      type: 'Amount'
      value?: Amount
      defaultValue?: Amount
      trailingZeros?: boolean
    }
  | {
      type: 'AmountFiat'
      value?: AmountFiat
      defaultValue?: AmountFiat
      trailingZeros?: boolean
    }
  | {
      type: 'AmountPercentage'
      value?: AmountPercentage
      defaultValue?: AmountPercentage
      trailingZeros?: boolean
      units?: never
    }
  | {
      type: 'AmountCrypto'
      value?: AmountCrypto
      defaultValue?: AmountCrypto
      trailingZeros?: never
    }
  | {
      type: 'AmountCurrency'
      value?: AmountCurrency
      defaultValue?: AmountCurrency
      trailingZeros?: boolean
    }

export type AmountValue = Amount | AmountPercentage | AmountFiat | AmountCrypto | AmountCurrency
