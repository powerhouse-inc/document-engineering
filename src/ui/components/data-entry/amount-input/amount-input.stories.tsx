import { commonCryptoCurrencies, commonFiatCurrencies } from '../currency-code-picker/defaults.js'
import type { Meta, StoryObj } from '@storybook/react'
import {
  getDefaultArgTypes,
  getValidationArgTypes,
  PrebuiltArgTypes,
  StorybookControlCategory,
} from '../../../../scalars/lib/storybook-arg-types.js'
import { AmountInput } from './index.js'

const mappedFiatCurrencies = commonFiatCurrencies.map((currency) => ({
  ...currency,
  label: currency.ticker,
}))
const mappedCryptoCurrencies = commonCryptoCurrencies.map((currency) => ({
  ...currency,
  label: currency.ticker,
}))

/**
 * The `AmountInput` component provides an input field for entering and selecting monetary amounts.
 * It supports multiple configuration properties like:
 * - label
 * - description
 * - units (currencies)
 * - precision
 * - step
 * - currency position
 *
 * Features include:
 * - Support for both fiat and crypto currencies
 * - Customizable decimal precision
 * - Currency symbol positioning
 * - Min/Max value restrictions
 * - Negative value support
 * - Trailing zeros option
 * - Multiple amount types (Fiat, Crypto, Percentage, etc.)
 *
 * > **Note:** This component does not have built-in validation. If you need built-in validation
 * > you can use the [AmountField](?path=/docs/scalars-amount-field--readme)
 * > component.
 */
const meta = {
  title: 'Data Entry/Amount Input',
  component: AmountInput,
  parameters: {
    layout: 'centered',
    form: {
      defaultValues: {
        'amount-field': {
          amount: undefined,
          unit: '',
        },
      },
    },
  },
  tags: ['autodocs'],
  argTypes: {
    units: {
      control: 'object',
      description: 'Array of Currency objects',
      table: {
        type: { summary: 'Currency[]' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
      if: {
        arg: 'type',
        neq: 'AmountPercentage',
      },
    },
    includeCurrencySymbols: {
      control: 'boolean',
      description: ' Whether to display currency symbols alongside codes',
      table: {
        defaultValue: { summary: 'true' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
      if: {
        arg: 'type',
        neq: 'AmountPercentage',
      },
    },
    symbolPosition: {
      control: 'select',
      description: 'Position of the currency symbol',
      options: ['left', 'right'],
      table: {
        type: { summary: 'string' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
      if: {
        arg: 'type',
        neq: 'AmountPercentage',
      },
    },
    step: {
      control: 'number',
      description: 'The step value for the amount field',
      table: {
        defaultValue: { summary: '1' },
        type: { summary: 'number' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    currencyPosition: {
      control: 'select',
      options: ['left', 'right'],
      description: 'Determines the position of the currency select dropdown relative to the amount input field.',
      table: {
        defaultValue: { summary: 'right' },
        type: { summary: 'string' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
      if: {
        arg: 'type',
        neq: 'AmountPercentage',
      },
    },

    trailingZeros: {
      control: 'boolean',
      description:
        'When precision is set, for example to 2, determines if the the trailing zeros should be preserved ( for example: 25.00,7.50, etc.) or not ( for example: 25, 7.5).',
      if: {
        arg: 'type',
        neq: 'AmountToken',
      },
      table: {
        type: { summary: 'boolean' },
        category: StorybookControlCategory.VALIDATION,
      },
    },
    viewPrecision: {
      control: 'number',
      description: 'Number of decimal places viewed',
      table: {
        type: { summary: 'number' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    precision: {
      control: 'number',
      description: 'Number of decimal places viewed',
      table: {
        type: { summary: 'number' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    allowNegative: {
      control: 'boolean',
      description: 'Whether negative values are allowed (true or false).',
      table: {
        defaultValue: { summary: 'false' },
        type: { summary: 'boolean' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    numberProps: {
      control: 'object',
      description: 'All the props options for number field',
      table: {
        type: { summary: 'string' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    selectProps: {
      control: 'object',
      description: 'All the props options for select field',
      table: {
        type: { summary: 'string' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    ...getDefaultArgTypes(),

    value: {
      control: 'object',
      description:
        "The value of the amount field. Can be a number, an object with currency, or undefined. Examples: { amount: 100, currency: 'USD' }, 200, undefined.",
      table: {
        type: { summary: 'object | number | undefined' },
        category: StorybookControlCategory.DEFAULT,
      },
    },
    ...PrebuiltArgTypes.placeholder,
    ...PrebuiltArgTypes.minValue,
    ...PrebuiltArgTypes.maxValue,
    ...PrebuiltArgTypes.viewMode,
    baseValue: {
      control: 'object',
      description: 'The base value of the amount field.',
      table: {
        type: { summary: 'Amount | AmountFiat | AmountPercentage | AmountCrypto | AmountCurrency' },
        category: StorybookControlCategory.DIFF,
      },
    },
    type: {
      control: 'select',
      options: ['Amount', 'AmountFiat', 'AmountPercentage', 'AmountCrypto', 'AmountCurrency'],
      description: 'The type of amount field.',
      table: {
        type: { summary: 'string' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    ...getValidationArgTypes({
      enabledArgTypes: {
        validators: false,
        showErrorOnBlur: false,
        showErrorOnChange: false,
      },
    }),
  },

  args: {
    name: 'amount-field',
  },
} satisfies Meta<typeof AmountInput>

export default meta
type Story = StoryObj<typeof AmountInput>

export const Default: Story = {
  args: {
    placeholder: '0',
    units: mappedFiatCurrencies,
    label: 'Enter Amount and Select Currency',
    placeholderSelect: 'CUR',
    type: 'Amount',
    name: 'amount-field',
    value: {
      amount: undefined,
      unit: '',
    },
  },
}

export const WithValue: Story = {
  args: {
    placeholder: 'Enter Amount',
    placeholderSelect: 'CUR',
    label: 'Enter Amount',
    type: 'Amount',
    value: {
      amount: 100,
    },
  },
}
export const WithAmount: Story = {
  parameters: {
    form: {
      defaultValues: {
        'amount-field': '',
      },
    },
  },
  args: {
    units: mappedFiatCurrencies,
    placeholder: 'Enter Amount',
    placeholderSelect: 'CUR',
    label: 'Enter Amout ',
    type: 'AmountFiat',
    name: 'amount-field',
    value: {
      amount: 345,
      unit: 'EUR',
    },
  },
}
export const CurrencyIcon: Story = {
  args: {
    units: mappedCryptoCurrencies,
    placeholder: 'Enter Amount',
    label: 'Enter Amount and Select Currency',
    type: 'AmountCrypto',
    placeholderSelect: 'CUR',
    value: {
      amount: 3454564564 as unknown as bigint,
      unit: 'BTC',
    },
  },
}

export const WithToken: Story = {
  parameters: {
    units: mappedCryptoCurrencies,
    form: {
      defaultValues: {
        'amount-field': {
          amount: '',
          unit: '',
        },
      },
    },
  },
  args: {
    placeholder: 'Enter Amount',
    label: 'Enter Amount and Select Currency',
    type: 'AmountCurrency',
    placeholderSelect: 'CUR',
    units: [...mappedCryptoCurrencies, ...mappedFiatCurrencies],
    value: {
      amount: 123 as unknown as bigint,
      unit: 'BTC',
    },
  },
}

export const WithValuePercent: Story = {
  parameters: {
    form: {
      defaultValues: {
        'amount-field': '',
      },
    },
  },
  args: {
    label: 'Enter Percentage ',
    placeholder: 'Enter Amount',
    type: 'AmountPercentage',
    value: 9,
  },
}
export const Disable: Story = {
  args: {
    units: mappedCryptoCurrencies,
    label: 'Enter Amount ',
    placeholder: 'Enter Amount',
    type: 'AmountFiat',
    placeholderSelect: 'CUR',
    disabled: true,
    value: {
      amount: 9,
      unit: 'ETH',
    },
  },
}

export const WithValueUniversalAmountCurrency: Story = {
  parameters: {
    form: {
      defaultValues: {
        'amount-field': {
          amount: 123,
          unit: 'BTC',
        },
      },
    },
  },
  args: {
    units: [...mappedCryptoCurrencies, ...mappedFiatCurrencies],
    label: 'Label',
    placeholder: 'Enter Amount',
    placeholderSelect: 'CUR',
    type: 'AmountCurrency',
    value: {
      amount: 123,
      unit: 'BTC',
    },
  },
}

export const WithDifferencesAddition: Story = {
  args: {
    type: 'AmountFiat',
    units: mappedFiatCurrencies,
    label: 'Icon names addition',
    value: {
      amount: 22233,
      unit: 'USD',
    },
    baseValue: {
      amount: 42323,
      unit: 'THB',
    },
    viewMode: 'addition',
  },
}

export const WithDifferencesRemoval: Story = {
  args: {
    label: 'Icon names removal',
    value: {
      amount: 12323,
      unit: 'ETH',
    },
    baseValue: {
      amount: 413130,
      unit: 'BTC',
    },
    type: 'AmountCurrency',
    units: [...mappedCryptoCurrencies, ...mappedFiatCurrencies],
    viewMode: 'removal',
  },
}

export const WithDifferencesMixed: Story = {
  args: {
    label: 'Icon names mixed',
    value: {
      amount: 909123223,
      unit: 'BTC',
    },
    baseValue: {
      amount: 49989822,
      unit: 'ETH',
    },
    units: [...mappedCryptoCurrencies, ...mappedFiatCurrencies],
    viewMode: 'mixed',
  },
}
