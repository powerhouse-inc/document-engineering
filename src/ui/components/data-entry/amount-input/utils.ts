import { cryptoCurrencies, currencies, fiatCurrencies } from '../currency-code-picker/utils.js'
import type { Currency } from '../currency-code-picker/types.js'
import type { AmountInputPropsGeneric, AmountValue } from './types.js'

export const getDefaultUnits = (type: AmountInputPropsGeneric['type']) => {
  switch (type) {
    case 'AmountFiat':
      return fiatCurrencies()
    case 'AmountCrypto':
      return cryptoCurrencies()
    case 'AmountCurrency':
      return currencies()
    case 'Amount':
    case 'AmountPercentage':
      return []
  }
}

/**
 * Checks if a given string value represents a valid BigInt.
 * A valid BigInt can be a positive or negative integer, or zero,
 * without decimal points or other non-digit characters (except for the leading minus sign).
 *
 * @param value The string to validate.
 * @returns True if the value is a valid BigInt string, false otherwise.
 */
export const isValidBigInt = (value: string | undefined): boolean => {
  // If the value is undefined or an empty string, it's not a valid BigInt.
  if (!value) {
    return false
  }

  const bigintRegex = /^-?\d+$/

  return bigintRegex.test(value)
}

export const isValidNumberGreaterThanMaxSafeInteger = (value: string | undefined) => {
  if (!value) {
    return false
  }
  const isValidBigIntValue = isValidBigInt(value)
  return isValidBigIntValue && Math.abs(Number(value)) > Number.MAX_SAFE_INTEGER
}

export const isNotSafeValue = (value: string) => {
  return Math.abs(Number(value)) > Number.MAX_SAFE_INTEGER
}

interface DisplayValueAmountProps {
  value?: string
  precision?: number
  viewPrecision?: number
  trailingZeros?: boolean
}

export const displayValueAmount = ({ value, precision, viewPrecision, trailingZeros }: DisplayValueAmountProps) => {
  if (!value) {
    return undefined // Return undefined if no value is provided
  }

  // If viewPrecision is provided but not precision, format to viewPrecision
  if (viewPrecision !== undefined && precision === undefined) {
    const formattedValue = parseFloat(value).toFixed(viewPrecision)
    return trailingZeros ? formattedValue : parseFloat(formattedValue).toString()
  }

  // If precision is provided but not viewPrecision, truncate to precision
  if (precision !== undefined && viewPrecision === undefined) {
    const formattedValue = parseFloat(value).toFixed(precision)

    // If the number of decimals is greater than or equal to the precision, apply toFixed
    return trailingZeros ? formattedValue : parseFloat(formattedValue).toString()
  }

  // If both are present, show the value with viewPrecision when focused
  if (precision !== undefined && viewPrecision !== undefined) {
    const formattedValue = parseFloat(value).toFixed(viewPrecision)
    return trailingZeros ? formattedValue : parseFloat(formattedValue)
  }
  return value
}

export const handleEventOnChange = <T>(value: T) => {
  const nativeEvent = new Event('change', { bubbles: true, cancelable: true })

  Object.defineProperty(nativeEvent, 'target', {
    value: { value },

    writable: false,
  })

  return nativeEvent as unknown as React.ChangeEvent<HTMLInputElement>
}

export const handleEventOnBlur = <T>(value: T) => {
  const nativeEvent = new Event('blur', { bubbles: true, cancelable: true })

  Object.defineProperty(nativeEvent, 'target', {
    value: { value },

    writable: false,
  })

  return nativeEvent as unknown as React.FocusEvent<HTMLInputElement>
}

export const createAmountValue = (inputValue: string) => {
  return inputValue === '' ? undefined : inputValue
}

export const isValidUnit = (type: AmountInputPropsGeneric['type'], value: AmountValue, units?: Currency[]) => {
  if (!units) return true
  if (type === 'Amount' && typeof value === 'object' && 'unit' in value) {
    return units.some((u) => u.ticker === value.unit)
  }
  if (type === 'AmountCurrency' && typeof value === 'object') {
    return units.some((u) => u.ticker === value.unit)
  }
  if (type === 'AmountCrypto' && typeof value === 'object') {
    return units.some((u) => u.ticker === value.unit)
  }
  if (type === 'AmountFiat' && typeof value === 'object') {
    return units.some((u) => u.ticker === value.unit)
  }
  return false
}
