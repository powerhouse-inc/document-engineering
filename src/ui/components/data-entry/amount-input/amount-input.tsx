import { forwardRef, useId, useState } from 'react'
import {
  FormDescription,
  FormGroup,
  FormLabel,
  FormMessageList,
} from '../../../../scalars/components/fragments/index.js'
import { cn } from '../../../../scalars/lib/utils.js'
import { type Currency, CurrencyCodePicker } from '../currency-code-picker/index.js'
import type { SelectProps } from '../select/index.js'
import type { Amount, AmountInputPropsGeneric, AmountValue } from './types.js'
import { useAmountInput } from './use-amount-input.js'
import { NumberInput } from '../number-input/index.js'
import type { InputNumberProps } from '../number-input/types.js'
import type { WithDifference } from '#scalars'
import { AmountInputDiff } from './subcomponents/amount-input-diff.js'

type AdditionalProps = Omit<InputNumberProps, 'onChange' | 'onBlur' | 'precision' | 'value' | 'defaultValue'> & {
  className?: string
  name: string
  numberProps?: Omit<InputNumberProps, 'name'>
  selectProps?: Omit<SelectProps, 'placeholder' | 'selectionIcon' | 'onBlur' | 'baseValue' | 'defaultValue' | 'value'>
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void
  onBlur?: (event: React.FocusEvent<HTMLInputElement>) => void
  currencyPosition?: 'left' | 'right'
  symbolPosition?: 'left' | 'right'
  allowNegative?: boolean
  viewPrecision?: number
  precision?: number
  placeholderSelect?: string
  units?: Currency[]
  includeCurrencySymbols?: boolean
}

interface AmountInputUncontrollerMain {
  type: 'Amount'
  value?: Amount
  defaultValue?: Amount
  trailingZeros?: boolean
}

type AmountInputUncontrollerProps = AmountInputUncontrollerMain & AdditionalProps & WithDifference<AmountValue>

type AmountInputProps = AmountInputPropsGeneric & AdditionalProps & WithDifference<AmountValue>

const AmountInputController = forwardRef<HTMLInputElement, AmountInputProps>(
  (
    {
      label,
      value,
      id: propId,
      minValue,
      maxValue,
      onChange,
      onBlur,
      disabled,
      className,
      required,
      errors,
      warnings,
      description,
      defaultValue,
      type,
      numberProps,
      selectProps,
      step = 1,
      currencyPosition = 'right',
      name,
      trailingZeros,
      viewPrecision,
      precision,
      placeholder,
      placeholderSelect,
      units,
      includeCurrencySymbols,
      symbolPosition,
      // Diff Props
      baseValue,
      viewMode = 'edition',
    },
    ref
  ) => {
    const generatedId = useId()
    const id = propId ?? generatedId
    const {
      isShowSelect,
      isPercent,
      options,
      valueSelect,
      valueInput,
      handleOnChangeInput,
      handleOnChangeSelect,
      handleBlur,
      isBigInt,
      handleIsInputFocused,
      isAmountWithoutUnit,
      inputFocused,
    } = useAmountInput({
      value,
      defaultValue,
      type,
      onChange,
      onBlur,
      precision,
      viewPrecision,
      trailingZeros,
      units,
    })

    if (viewMode === 'edition') {
      return (
        <FormGroup>
          {label && (
            <FormLabel
              htmlFor={id}
              required={required}
              disabled={disabled}
              hasError={!!errors?.length}
              className={cn(disabled && 'mb-[3px]')}
            >
              {label}
            </FormLabel>
          )}
          <div
            className={cn(
              'relative flex items-center',
              'rounded-md border border-gray-300',
              'h-[36px]',
              // focus state
              'focus:rounded-md focus:border-gray-900 transition-colors focus-within:ring-1 focus-within:rounded focus-within:ring-offset-0'
            )}
          >
            <input name={name} type="hidden" data-cast={isBigInt ? 'AmountBigInt' : 'AmountNumber'} />
            {isShowSelect && currencyPosition === 'left' && (
              <CurrencyCodePicker
                contentAlign="start"
                contentClassName="[&]:!w-[120px]"
                disabled={disabled}
                currencies={options}
                value={valueSelect}
                onChange={handleOnChangeSelect}
                placeholder={placeholderSelect}
                includeCurrencySymbols={includeCurrencySymbols}
                symbolPosition={symbolPosition}
                searchable={false}
                className={cn(
                  'focus:border-r-none focus:z-10 focus:ring-0 focus:ring-offset-0 focus:bg-none',
                  'outline-none focus-visible:ring-0 focus-visible:ring-offset-0',
                  'h-[32px]',
                  'hover:bg-transparent',
                  'border-none',
                  selectProps?.className
                )}
                {...(selectProps ?? { name: '' })}
              />
            )}
            {isShowSelect && currencyPosition === 'left' && (
              <div className="border-l-[1px] border-gray-300 h-full flex items-center" />
            )}
            <div className={cn('relative flex items-center')}>
              <NumberInput
                name=""
                step={step}
                required={required}
                disabled={disabled}
                value={valueInput === undefined ? undefined : (valueInput as unknown as number)}
                id={id}
                maxValue={maxValue}
                precision={precision}
                minValue={minValue}
                onChange={handleOnChangeInput}
                onFocus={handleIsInputFocused}
                placeholder={placeholder}
                numericType={type === 'AmountCurrency' || type === 'AmountCrypto' ? 'BigInt' : 'Float'}
                className={cn(
                  currencyPosition === 'left' && 'rounded-l-none border-l-[0.5px]',
                  currencyPosition === 'right' && 'rounded-r-none border-r-[0.5px]',
                  isPercent && 'rounded-md pr-7',
                  isAmountWithoutUnit && 'rounded-md',
                  'outline-none focus-visible:ring-0 focus-visible:ring-offset-0',
                  'border-none',
                  'h-[32px]',
                  className
                )}
                onBlur={handleBlur}
                ref={ref}
                {...(numberProps || {})}
              />
              {isPercent && !inputFocused && (
                <span
                  className={cn(
                    'pointer-events-none absolute inset-y-0 right-2 ml-2 flex items-center',
                    disabled ? 'text-gray-400' : 'text-gray-900'
                  )}
                >
                  %
                </span>
              )}
            </div>
            {isShowSelect && currencyPosition === 'right' && (
              <div className="border-l-[1px] border-gray-300 h-full flex items-center" />
            )}
            {isShowSelect && currencyPosition === 'right' && (
              <CurrencyCodePicker
                contentAlign="end"
                contentClassName="[&]:!w-[120px]"
                className={cn(
                  'focus:border-l-none focus:z-10 focus:ring-0 focus:ring-offset-0 focus:bg-none',
                  'outline-none focus-visible:ring-0 focus-visible:ring-offset-0',
                  'h-[32px]',
                  'hover:bg-transparent',
                  'border-none',
                  selectProps?.className
                )}
                disabled={disabled}
                includeCurrencySymbols={includeCurrencySymbols}
                currencies={options}
                value={valueSelect}
                onChange={handleOnChangeSelect}
                name=""
                placeholder={placeholderSelect}
                symbolPosition={symbolPosition}
                searchable={false}
                {...(selectProps ?? { name: '' })}
              />
            )}
          </div>
          {description && <FormDescription>{description}</FormDescription>}
          {warnings && <FormMessageList messages={warnings} type="warning" />}
          {errors && <FormMessageList messages={errors} type="error" />}
        </FormGroup>
      )
    }
    return (
      <AmountInputDiff
        value={value}
        baseValue={baseValue}
        viewMode={viewMode}
        required={required}
        label={label}
        currencyPosition={currencyPosition}
        options={options}
        symbolPosition={symbolPosition}
        includeCurrencySymbols={includeCurrencySymbols}
      />
    )
  }
)
AmountInputController.displayName = 'AmountInputController'

const AmountInputUncontroller = (props: AmountInputUncontrollerProps) => {
  const [value, setValue] = useState(props.value ?? props.defaultValue)

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = e.target.value
    setValue(newValue as Amount)
  }
  return <AmountInputController {...props} value={value} onChange={handleChange} />
}

const AmountInput = (props: AmountInputProps) => {
  if (props.onChange) {
    return <AmountInputController {...props} />
  }
  return <AmountInputUncontroller {...(props as AmountInputUncontrollerProps)} />
}

AmountInput.displayName = 'AmountInput'

export { AmountInput, type AmountInputProps }
