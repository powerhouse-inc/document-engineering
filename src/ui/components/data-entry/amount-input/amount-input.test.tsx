import { render, screen } from '@testing-library/react'
import { AmountInput } from './amount-input.js'
import { commonCryptoCurrencies } from '../currency-code-picker/defaults.js'

describe('AmountField Component', () => {
  const onChange = vi.fn()
  it('should match snapshot', () => {
    const { container } = render(
      <AmountInput
        label="Amount Label"
        name="amount"
        type="Amount"
        value={{
          amount: 345,
        }}
        units={commonCryptoCurrencies}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('should render label when provided', () => {
    render(
      <AmountInput
        label="Amount Label"
        name="amount"
        type="Amount"
        value={{
          amount: 345,
        }}
        units={commonCryptoCurrencies}
      />
    )
    expect(screen.getByLabelText('Amount Label')).toBeInTheDocument()
  })

  it('should render the percentage sign if the type is percent', () => {
    render(<AmountInput label="Amount Label" name="amount" type="AmountPercentage" value={345} step={0} />)
    expect(screen.getByText('%')).toBeInTheDocument()
  })

  it('should disable the input when disabled prop is true', () => {
    render(
      <AmountInput
        label="Amount Label"
        name="amount"
        type="Amount"
        value={{
          amount: 345,
        }}
        units={commonCryptoCurrencies}
        disabled
      />
    )
    const input = screen.getByLabelText('Amount Label')

    expect(input).toBeDisabled()
  })

  it('should set the input as required when required prop is true', () => {
    render(
      <AmountInput
        label="Amount Label"
        name="amount"
        type="Amount"
        value={{
          amount: 345,
        }}
        required
        units={commonCryptoCurrencies}
      />
    )
    expect(screen.getByRole('spinbutton')).toHaveAttribute('required')
  })

  it('should render the description when provided', () => {
    render(
      <AmountInput
        label="Amount Label"
        name="amount"
        type="Amount"
        value={{
          amount: 345,
        }}
        units={commonCryptoCurrencies}
        disabled
        description="This is a description"
      />
    )
    expect(screen.getByText('This is a description')).toBeInTheDocument()
  })

  describe('AmountInputDiff differences', () => {
    it('should render AmountInputDiff when viewMode is addition', () => {
      render(
        <AmountInput
          type="Amount"
          name="amount"
          viewMode="addition"
          onChange={onChange}
          value={{
            amount: 345,
          }}
          baseValue={{
            amount: 1345,
          }}
        />
      )
      expect(screen.getByTestId('amount-input-diff')).toBeInTheDocument()
      expect(screen.getByText('345')).toBeInTheDocument()
      expect(screen.queryByText('1345')).not.toBeInTheDocument()
    })

    it('should render DateInputDiff when viewMode is removal', () => {
      render(
        <AmountInput
          name="amount"
          type="AmountCurrency"
          viewMode="removal"
          value={{
            amount: '9345',
            unit: 'USD',
          }}
          baseValue={{
            amount: 2345,
            unit: 'USD',
          }}
        />
      )
      expect(screen.getByTestId('amount-input-diff')).toBeInTheDocument()
      expect(screen.getByText('2345')).toBeInTheDocument()
      expect(screen.queryByText('9345')).not.toBeInTheDocument()
    })

    it('should render DateInputDiff when viewMode is mixed', () => {
      render(
        <AmountInput
          name="amount"
          type="AmountCurrency"
          viewMode="mixed"
          value={{
            amount: '2345',
            unit: 'USD',
          }}
          baseValue={{
            amount: 1345,
            unit: 'USD',
          }}
        />
      )
      expect(screen.getByTestId('amount-input-diff')).toBeInTheDocument()
      expect(screen.getByText('1345')).toBeInTheDocument()
      expect(screen.getByText('2345')).toBeInTheDocument()
    })

    it('should not render DateInputDiff when viewMode is edition', () => {
      render(
        <AmountInput
          type="AmountCurrency"
          name="amount"
          viewMode="edition"
          value={{
            amount: '345',
            unit: 'USD',
          }}
          baseValue={{
            amount: 345,
            unit: 'USD',
          }}
        />
      )
      expect(screen.queryByTestId('amount-input-diff')).not.toBeInTheDocument()
    })
  })
})
