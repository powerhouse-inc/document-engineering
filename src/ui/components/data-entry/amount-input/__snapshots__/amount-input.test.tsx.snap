// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AmountField Component > should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <label
      class="inline-flex items-center text-sm font-medium leading-4 text-gray-900 dark:text-gray-50 mb-[3px]"
      for=":r0:"
    >
      Amount Label
    </label>
    <div
      class="relative flex items-center rounded-md border border-gray-300 h-[36px] focus:rounded-md focus:border-gray-900 transition-colors focus-within:ring-1 focus-within:rounded focus-within:ring-offset-0"
    >
      <input
        data-cast="AmountNumber"
        name="amount"
        type="hidden"
      />
      <div
        class="relative flex items-center"
      >
        <div
          class="flex flex-col gap-2"
        >
          <div
            class="relative flex items-center group"
          >
            <input
              aria-invalid="false"
              class="flex w-full text-sm font-normal leading-5 text-gray-900 dark:text-gray-50 dark:border-charcoal-700 dark:bg-charcoal-900 border border-gray-300 bg-white px-3 py-2 font-sans placeholder:text-gray-500 dark:placeholder:text-gray-600 focus-visible:outline-none focus-visible:ring-gray-900 focus-visible:ring-offset-white dark:focus-visible:ring-charcoal-300 dark:focus-visible:ring-offset-charcoal-900 dark:focus:bg-charcoal-900 focus:bg-gray-50 disabled:cursor-not-allowed disabled:border-gray-300 disabled:bg-gray-50 disabled:text-gray-700 disabled:dark:border-charcoal-800 disabled:dark:bg-charcoal-900 disabled:dark:text-gray-300 pr-8 border-r-[0.5px] rounded-md outline-none focus-visible:ring-0 focus-visible:ring-offset-0 border-none h-[32px]"
              data-cast="Number"
              id=":r0:"
              inputmode="numeric"
              name=""
              role="spinbutton"
              type="text"
              value="345"
            />
            <div
              class="absolute inset-y-2 right-3 flex flex-col justify-center opacity-0 group-focus-within:opacity-100 transition-opacity"
            >
              <button
                aria-label="Increment"
                type="button"
              >
                <svg
                  class="rotate-180 text-gray-700 dark:text-gray-300"
                  data-testid="icon-fallback"
                  fill="none"
                  stroke="currentcolor"
                  style="width: 10px; height: 10px;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 9L12 15L18 9"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                </svg>
              </button>
              <button
                aria-label="Decrement"
                type="button"
              >
                <svg
                  class="items-center justify-center text-gray-700 dark:text-gray-300"
                  data-testid="icon-fallback"
                  fill="none"
                  stroke="currentcolor"
                  style="width: 10px; height: 10px;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 9L12 15L18 9"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
